### PR TITLE
RFC: Cleanup case constructs

### PR DESCRIPTION
**Pull request by darkstar**
Question: would you be willing to accept cleanups of those huge switch constructs in a form like this? I could go through the source and do it if you want.

These are in some source files and are very hard to follow. IMHO a table-driven structure like what I did here is more maintainable.

OTOH, since this pattern repeats in a few source files, would it make sense to have a generic helper function that translates (cylinders,heads,sectors-per-track,sector-size)->MediaType and maybe (filesize)->MediaType?

### CODE REVIEW

**Comment by darkstar** [comment]
I could do some other types of cleanups too, and you could cherry-pick those commits that you want and ignore the rest. For example removing unused "using" directives, simplifying the "ImageInfo" (and other structs) initialization, cleaning deeply nested "if-else" constructs (I'm looking at you, BlindWrite5.cs ;-), simplifying Regexes, etc.

**Comment by claunia** [comment]
yes but NEVER EVER use var. really, NEVER xD.

var myint = 2; // PR rejected
Int32 myint = 2; // PR rejected
int myint = 2; // OK

Right now I'm doing that kind of refactoring using Rider (aka Resharper) so let's see what escapes it before. Can I notify you when I'm done someplace?

**Comment by claunia** [comment]
Also about that "chs->mediatype" is not a bad idea if you can solve how to differentiate:
Amiga from Atari: 80/2/11/512
NEC MD2HD from MF2HD: 77/2/8/1024
and several others.

I think that part best leave it after feature freeze: Refactor as a table that's private to each disk image OK, a general method outside disk image plugins AFTER refactor (approx. >25th Dec)

**Comment by darkstar** [comment]
darn, I failed the PR, had to rebase. But it seems you understood what I intended  ;-)

And regarding "var", I don't really see any reason why not to use it, especially since it's the only way you can do anonymous structs in C#... If you have to declare the MediaInfoMap structure in each source file it doesn't really help simplify the code since the code will certainly be longer than before.
Note that I only suggest the use of "var" in cases where no other solution is possible to keep the code simple, I would never suggest "var myint = 2;" or something silly like this. I think in this case here, "var" is the cleanest solution.

Also, for Regexes, I became a big fan of the `@""` syntax, which lets you use a single backslash in a string literal... But I understand that this might also be a bit controversial

**Comment by darkstar** [comment]
And yes, I thought about the ambiguities in the chs->mediatype conversion too, and it might indeed be a problem. However, usually, each ImagePlugin has a fairly certain view of what type of floppies it could include and what not, so maybe the generic map function could return a list of possible matches, which the ImagePlugin then could filter according to platforms or whatnot, so the ADF plugin would filter out the Atari MediaType and the Atari plugin would reject the Amiga MediaType. But of course this is a bit more complex  and would better be tracked as a GitHub issue or in a separate branch for a while...

**Comment by claunia** [comment]
there are some things in DiscImageChef that are done C-style and not OOP-style. You know I didn't chose C-life, C-life chose me.

Seriously if there is absolutely no other way of doing a thing better than using var, I'll explore that in a case-by-case but in this particular PR, is not that case.

For <code>@""</code> I have nothing against that syntax, just I'm not used to use it, but that refactoring is accepted.

**Comment by darkstar** [comment]
Hey, even C++ has "auto" now and people seem to absolutely love it... Back in my day, people were proud of writing `std::map<std::pair<int, std::string>, std::list<std::string>>::iterator it = mymap.begin();`, now everyone is just lazy and writes `auto it = mymap.begin();` :-D

Aaaaanyway, right now I don't see how that particular local table could be done differently without having to declare a separate struct with a constructor, but I admit I don't know all the nooks and crannies of C#... Maybe there's something like a `tuple<int, int, int, MediaType>` that could be used?

**Comment by claunia** [comment]
well what you put there much seems like a struct constructor  to me:

```
struct MediaTypeCharacteristics
{
 uint cylinders;
 byte heads;
 ushort sectorsPerTrack;
 ushort bytesPerSector;
 MediaType mediaType
}

MediaTypeCharacteristics mtTable[] = new {
{ cylinders = 80, heads = 2, sectorsPerTrack = 9, bytesPerSector = 512, mediaType = MediaType.DOS_35_DD_9 } };
```

Tuples are another way to do it and you can always do:
`using MediaTypeCharacteristics = Tuple<uint, byte, ushort, ushort, MediaType>;`

**Comment by darkstar** [comment]
Yes, indeed it is a constructor, but for an anonymous type... I didn't realize that the same assignment (the `new { foo="bar", baz=23 }` style one) should also work for non-anonymous types...

**Comment by claunia** [comment]
Basically an anonymous struct type is no different from a named one in the framework. The problem is that when you use an anonymous one the compiler choices the type and order of the fields, and that's a choice it's better done by us, with corresponding XML documentation "this is that, this is thet, this is thot".

Maybe it's my heritage but int32, int32, int32, int32, when you are safe using uint32, uint8, uint16, uint16 seems to be a misuse of memory 😝

**Comment by claunia** [comment]
Superseeded by 741a3289a24fd4df2c29e7ba97bd4206749d4509.
